### PR TITLE
Expose formation ID

### PR DESCRIPTION
**Pull request by mattetre**
- expose current formation `id` on Generator objects
- add an `onProcess` that can be overridden to do some handling when `process` runs (for example check the current formation ID to see if we are finished)

### CODE REVIEW

File: lib/Generator.ts [hunk 1]
```diff
@@ -149,6 +158,13 @@ export class Generator<Data> extends ChatBotGenerator<IParameters, Data> {
    */
   protected convertSuggestion?(suggestion: Data): IInputImmutable;
 
+  /**
+   * This will be called when the iterator finishes processing a piece of discourse
+   * Override this if you want to do some handling after discource has been processed
+   * @param discourse the discourse that was processed
+   */
+  protected async onProcess?(discourse: IDiscourseImmutable<Data>): Promise<void>;
```
Comment (mattyclarkson) [inline lib/Generator.ts]: Usually the derived classes just override the base implementation in class hierarchies. You could make the `process` function protected and call super in derived classes:

```typescript
class MyGenerator extends Generator<string> {
  protected async onProcess(discourse: IDiscourseImmutable<Data>): Promise<void> {
    // My stuff
  }
}
```

```typescript
class MyGenerator extends Generator<string> {
  protected async process(discourse: IDiscourseImmutable<Data>): Promise<void> {
    super(discourse);
    // My stuff
  }
}
```

The advantage (maybe disadvantage) you get is that you can then customise the process completely, if you wanted too.

File: lib/Generator.ts [hunk 2]
```diff
@@ -39,19 +39,24 @@ export interface IParameters {
...
+    this.id = new Guid(id);
     this.context = context ? { ...context } : {};
     this.classifier = classifyConvert(classifier);
 
     const descriptors = Object.getOwnPropertyDescriptors(this);
     descriptors.walker.writable = false;
     descriptors.walker.configurable = false;
     descriptors.walker.enumerable = false;
+    descriptors.id.configurable = false;
+    descriptors.id.writable = true;
```
Comment (mattyclarkson) [inline lib/Generator.ts]: Ideally this should be `readonly`, so people don't think that setting the `id` will actually do anything. It's a bit of a faff to make it `readonly` though as you want to hide the `setter` from derived classes. It is possible though. This is the *private data* idiom. This stores the immutable formation (as I'm sure something later will want to get currently processed formation).

*Private Data Interface*
```typescript
interface IData<Data> {
  formation: IFormationImmutable<Data> | undefined;
}
```

*Member variable:*
```typescript
protected readonly formation!: IFormationImmutable<Data> | undefined;
private readonly setFormation: (formation: IFormationImmutable<Data>) => void;
```

*Constructor:*
```typescript
const data = { formation: undefined };
Object.defineProperties(this, {
  setFormation: { value: (formation: IFormationImmutable<Data>): void => { data.formation = formation; } },
  formation: { get: () => data.formation },
});
```

*Process:*
```typescript
  this.setFormation(value);  // Probably makes sense to remap { done, value: formation }
```

*ID accessor*:
```typescript
get id(): IGuidImmutable {
  const formation = this.formation;
  return formation ? formation.id : this.walker.start;
}
```

Comment (mattyclarkson) [reply lib/Generator.ts]: Ignore all that 😊 Use the new `this.walker.id` in `@ef-carbon/conversation` `2.8.0`. Do `yarn upgrade @ef-carbon/conversation` to get that. The just add the `id` accessor:

```typescript
get id(): IGuidImmutable {
  return this.walker.id;
}
```

**Comment by mattetre** [comment]
@mattyclarkson changes made. muuch simpler PR now 😄

**Pull request merged**